python/lsst/meas/astrom/matchOptimisticB.py

```diff
@@ -0,0 +1,241 @@
+            numCleanSources = numCleanSources,
+            minMatchedPairs = minMatchedPairs,
+            verbose = debug.verbose,
+        )
+        if matchList0 is not None:
+            matchList0 = [m for m in matchList0 
+                           if (not m.second.getCentroidFlag() and 
+                               not m.second.get("base_PixelFlags_flag_saturated"))]
+        else:
+            matchList0 = []
+
+        # match good (unsaturated) sources
+        if len(refCat) > len(cleanSourceCat) - len(goodSources):
+            matchList1 = self.doMatch(
+                refCat = refCat,
+                sourceCat = goodSources,
+                wcs = wcs,
+                refFluxField = refFluxField,
+                numCleanSources = numCleanSources,
+                minMatchedPairs = minMatchedPairs,
+                verbose = debug.verbose,
+            )
+        else:
+            matchList1 = []
+        if matchList1 == None:
+            matchList1 = []
+
+        if len(matchList0) == 0 and len(matchList1) == 0:
+            raise RuntimeError("Unable to match sources")
+
+        # Adopt matchList with more matches
+        if len(matchList0) > len(matchList1):
+            matchList = matchList0
+        else:
+            matchList = matchList1
+
+        if self.log: self.log.info("Matched %d sources" % len(matchList))
+        if len(matchList) < minMatchedPairs:
+            self.log.warn("Number of matches is smaller than request")
+
+        return pipeBase.Struct(
+            matchList = matchList,
+        )
+
+    @pipeBase.timeMethod
+    def doMatch(self, refCat, sourceCat, wcs, refFluxField, numCleanSources, minMatchedPairs, verbose):
+        """!Match sources to position reference stars
+
+        @param[in] refCat  catalog of position reference stars that overlap an exposure
+        @param[in] sourceCat  catalog of sources found on the exposure
+        @param[in] wcs  estimated WCS of exposure
+        @param[in] refFluxField  field of refCat to use for flux
+        @param[in] numCleanSources  number of clean sources (sources with known centroid
+            that are not near the edge, but may be saturated)
+        @param[in] minMatchedPairs  minimum number of matches
+        @param[in] verbose  true to print diagnostic information to std::cout
+        """
+        numSources = len(sourceCat)
+        posRefBegInd = numCleanSources - numSources
+        configMatchDistPix = self.config.maxMatchDistArcSec/wcs.pixelScale().asArcseconds()
+
+        matchControl = MatchOptimisticBControl()
+        matchControl.refFluxField = refFluxField
+        matchControl.numBrightStars = self.config.numBrightStars
+        matchControl.minMatchedPairs = self.config.minMatchedPairs
+        matchControl.maxOffsetPix = self.config.maxOffsetPix
+        matchControl.numPointsForShape = self.config.numPointsForShape
+        matchControl.maxDeterminant = self.config.maxDeterminant
```

Comment:
Ah, sorry I never replied to this before: you can actually do this with the stuff in pex_config as well: just don't use `LSST_CONTROL_FIELD` to define the control class data members that shouldn't be converted into config fields (i.e. just make them regular public data members).  The `LSST_CONTROL_FIELD` macro's only purpose is to provide special static methods that the Python config-generator code can use to create the Python-side Field object, so it's a little confusing to see it used at all when that's not being used.